crates/sui-cost/src/estimator.rs

```diff
@@ -28,48 +42,110 @@ pub enum CommonTransactionCosts {
     SharedCounterIncrement,
 }
 
-pub fn estimate_computational_costs_for_transaction(
-    tx_kind: TransactionKind,
-) -> Result<GasCostSummary, anyhow::Error> {
-    let cost_map = read_estimate_file()?;
-    let unsupported_tx_kind = Err(anyhow!("Transaction kind not supported for estimator yet"));
-    match tx_kind {
-        TransactionKind::Single(s) => match s {
-            SingleTransactionKind::TransferSui(t) => Ok(if t.amount.is_none() {
-                cost_map.get(&CommonTransactionCosts::TransferWholeSuiCoin)
-            } else {
-                cost_map.get(&CommonTransactionCosts::TransferPortionSuiCoin)
-            }
-            .unwrap()
-            .clone()),
-            SingleTransactionKind::TransferObject(_) => unsupported_tx_kind,
-            SingleTransactionKind::Pay(_) => unsupported_tx_kind,
-            SingleTransactionKind::Publish(_) => unsupported_tx_kind,
-            SingleTransactionKind::Call(_) => unsupported_tx_kind,
-            SingleTransactionKind::ChangeEpoch(_) => unsupported_tx_kind,
-        },
-        TransactionKind::Batch(_) => Err(anyhow!("Batch TXes not supported for estimator")),
+impl CommonTransactionCosts {
+    pub fn is_shared_object_tx(&self) -> bool {
+        matches!(
+            self,
+            CommonTransactionCosts::SharedCounterAssertValue
+                | CommonTransactionCosts::SharedCounterIncrement
+        )
     }
 }
 
-pub fn read_estimate_file(
-) -> Result<BTreeMap<CommonTransactionCosts, GasCostSummary>, anyhow::Error> {
-    let json_str = fs::read_to_string(ESTIMATE_FILE).unwrap();
+// Emulation-based estimator
+// Step 1: charge min tx fee : can be computed precisely
+// Step 2: if contains shared objs, charge consensus fee : can be computed precisely
+// Step 3: charge storage read for all input objects : can be computed precisely
+// Step 4: if package publish, charge for it per size of modules : can be computed precisely
+// Step 5: charge VM flat fee if uses VM : can be computed precisely
+// Step 6: charge for mutations, deletions, rebates: cannot be computed precisely, will approx
+fn estimate_transaction_inner<S>(
+    tx: TransactionKind,
+    computation_gas_unit_price: u64,
+    storage_gas_unit_price: u64,
+    mutated_object_sizes_after: Option<usize>,
+    storage_rebate: SuiGas,
```

Comment:
this seems similar to `mutated_object_sizes_after`, maybe we should make it optional and approx when necessary?

Reply:
`storage_rebate`? Yes I can make it optional